File: Python/String/D15_AddBinaryStrings.py
def add_binary(s1, s2):
    i, j = len(s1) - 1, len(s2) - 1
    carry = 0
    res = []

    while i >= 0 or j >= 0 or carry:
        total = carry
        if i >= 0:
            total += int(s1[i])
            i -= 1
        if j >= 0:
            total += int(s2[j])
            j -= 1

        res.append(str(total % 2))
        carry = total // 2

    result = ''.join(res[::-1]).lstrip('0')
    return result if result else '0'

# Example calls
print(add_binary("1101", "111"))   # Output: 10100
print(add_binary("00100", "010"))  # Output: 1110

File: CPP/String/D15_AddBinaryStrings.cpp
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;

string addBinary(string s1, string s2)
{
      int i = s1.length() - 1, j = s2.length() - 1, carry = 0;
      string res = "";

      while (i >= 0 || j >= 0 || carry)
      {
            int sum = carry;
            if (i >= 0)
                  sum += s1[i--] - '0';
            if (j >= 0)
                  sum += s2[j--] - '0';
            res += (sum % 2 + '0');
            carry = sum / 2;
      }

      reverse(res.begin(), res.end());

      int index = 0;
      while (index < res.length() && res[index] == '0')
      {
            index++;
      }

      return (index == res.length()) ? "0" : res.substr(index);
}
int main()
{
      string s1 = "1101", s2 = "111";
      string s3 = "00100", s4 = "010";

      cout << addBinary(s1, s2) << endl; 
      cout << addBinary(s3, s4) << endl; 
      return 0;
}
